src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,715 @@
+      urem = eq[1];
+      Assert(is_bv_const(eq[0]));
+      eqrhs = eq[0];
+    }
+    if (getMinBwExpr(Rewriter::rewrite(urem[0])) == 0)
+    {
+      Trace("bv-gauss-elim")
+          << "Minimum required bit-width exceeds given bit-width, "
+             "will not apply Gaussian Elimination."
+          << endl;
+      return BVGaussElim::Result::INVALID;
+    }
+    rhs.push_back(get_bv_const_value(eqrhs));
+
+    Assert(is_bv_const(urem[1]));
+    Assert(i == 0 || get_bv_const_value(urem[1]) == iprime);
+    if (i == 0)
+    {
+      prime = urem[1];
+      iprime = get_bv_const_value(prime);
+    }
+
+    unordered_map<Node, Integer, NodeHashFunction> tmp;
+    stack<Node> stack;
+    stack.push(urem[0]);
+    while (!stack.empty())
+    {
+      Node n = stack.top();
+      stack.pop();
+
+      /* Subtract from rhs if const */
+      if (is_bv_const(n))
+      {
+        Integer val = get_bv_const_value(n);
+        if (val > 0) rhs.back() -= val;
+        continue;
+      }
+
+      /* Split into matrix columns */
+      Kind k = n.getKind();
+      if (k == kind::BITVECTOR_PLUS)
+      {
+        for (const Node& nn : n) { stack.push(nn); }
+      }
+      else if (k == kind::BITVECTOR_MULT)
+      {
+        Node n0, n1;
+        /* Flatten mult expression. */
+        n = RewriteRule<FlattenAssocCommut>::run<true>(n);
+        /* Split operands into consts and non-consts */
+        NodeBuilder<> nb_consts(NodeManager::currentNM(), k);
+        NodeBuilder<> nb_nonconsts(NodeManager::currentNM(), k);
+        for (const Node& nn : n)
+        {
+          Node nnrw = Rewriter::rewrite(nn);
+          if (is_bv_const(nnrw))
+          {
+            nb_consts << nnrw;
+          }
+          else
+          {
+            nb_nonconsts << nnrw;
+          }
+        }
+        Assert(nb_nonconsts.getNumChildren() > 0);
+        /* n0 is const */
+        unsigned nc = nb_consts.getNumChildren();
+        if (nc > 1)
+        {
+          n0 = Rewriter::rewrite(nb_consts.constructNode());
+        }
+        else if (nc == 1)
+        {
+          n0 = nb_consts[0];
+        }
+        else
+        {
+          n0 = utils::mkOne(utils::getSize(n));
+        }
+        /* n1 is a mult with non-const operands */
+        if (nb_nonconsts.getNumChildren() > 1)
+        {
+          n1 = Rewriter::rewrite(nb_nonconsts.constructNode());
+        }
+        else
+        {
+          n1 = nb_nonconsts[0];
+        }
+        Assert(is_bv_const(n0));
+        Assert(!is_bv_const(n1));
+        tmp[n1] += get_bv_const_value(n0);
+      }
+      else
+      {
+        tmp[n] += Integer(1);
+      }
+    }
+
+    /* Note: "var" is not necessarily a VARIABLE but can be an arbitrary expr */
+
+    for (const auto& p : tmp)
+    {
+      Node var = p.first;
+      Integer val = p.second;
+      if (i > 0 && vars.find(var) == vars.end())
+      {
+        for (size_t j = 0; j < i; ++j) { vars[var].push_back(Integer(0)); }
```

Comment:
I just realized that this could also be written as `vars[var].insert(vars[var].end(), i, Integer(0));`.